File: kalba-rest-api/src/main/java/kalba/service/AccountService.java
package kalba.service;

import kalba.common.exception.AccountConflictException;
import kalba.models.account.Account;
import kalba.models.account.AccountInfo;
import kalba.repository.AccountRepository;
import lombok.AllArgsConstructor;

import javax.transaction.Transactional;
import java.util.List;
import java.util.Optional;

@Transactional
@AllArgsConstructor
public class AccountService {
    private final AccountRepository accountRepository;

    public int register(Account account) {
        if (validateDuplicateName(account) || validateDuplicateTag(account)) {
            throw new AccountConflictException(); // 중복 계정 이름 있음, 중복 태그 있음
        }

        Account registeredAccount = accountRepository.register(account);
        return registeredAccount.getId();
    }

    private boolean validateDuplicateName(Account account) {
        return accountRepository.findByName(account.getName()).isPresent();
    }

    private boolean validateDuplicateTag(Account account) {
        return accountRepository.findByTag(account.getTag()).isPresent();
    }

    public List<Account> findAllAccounts() {
        return accountRepository.findAll();
    }

    public List<AccountInfo> findAllAccountInfo() {
        return accountRepository.findAllAccountInfo();
    }

    public Optional<Account> findAccountByName(String name) {
        return accountRepository.findByName(name);
    }

    public Optional<AccountInfo> findAccountInfoByName(String name) {
        return accountRepository.findAccountInfoByName(name);
    }
}

File: kalba/src/main/kotlin/kr/kalba/application/AccountService.kt
package kr.kalba.application

import kr.kalba.domain.mongo.Account
import kr.kalba.infrastructure.constant.Errors
import kr.kalba.infrastructure.exception.CommonException
import kr.kalba.infrastructure.external.coc.ClashOfClanService
import kr.kalba.infrastructure.external.coc.dto.VerifyTokenResponse
import kr.kalba.infrastructure.external.security.JwtTokenUtil
import kr.kalba.infrastructure.external.security.JwtUserDetailsService
import kr.kalba.infrastructure.repository.AccountRepository
import kr.kalba.presentation.dto.LoginDto
import kr.kalba.presentation.dto.RegisterDto
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.authentication.BadCredentialsException
import org.springframework.security.authentication.DisabledException
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.util.ObjectUtils

@Service
class AccountService(
    val accountRepository: AccountRepository,
    val passwordEncoder: PasswordEncoder,
    val authenticationManager: AuthenticationManager,
    val jwtUserDetailsService: JwtUserDetailsService,
    val jwtTokenUtil: JwtTokenUtil,
    val clashOfClanService: ClashOfClanService
) {
    fun register(registerDto: RegisterDto.Request): Account {
        if (validateDuplicateName(registerDto)) {
            throw CommonException(Errors.REGISTER_DUPLICATE_NAME)
        } else if (validateDuplicateTag(registerDto)) {
            throw CommonException(Errors.REGISTER_DUPLICATE_TAG)
        }
        val encryptedPassword = passwordEncoder.encode(registerDto.password)
        return accountRepository.save(Account.of(registerDto, encryptedPassword))
    }

    fun verifyToken(tag: String, token: String): VerifyTokenResponse {
        return clashOfClanService.verifyToken(tag, token)
    }

    private fun validateDuplicateName(registerDto: RegisterDto.Request): Boolean {
        return !ObjectUtils.isEmpty(accountRepository.findByAccountName(registerDto.accountName))
    }

    private fun validateDuplicateTag(registerDto: RegisterDto.Request): Boolean {
        return !ObjectUtils.isEmpty(accountRepository.findByTag(registerDto.tag))
    }

    fun login(loginDto: LoginDto.Request): String {
        try {
            authenticationManager.authenticate(
                UsernamePasswordAuthenticationToken(
                    loginDto.accountName,
                    loginDto.password
                )
            )
        } catch (e: DisabledException) {
            throw CommonException(Errors.LOGIN_DISABLE_USER)
        } catch (e: BadCredentialsException) {
            throw CommonException(Errors.LOGIN_INVALID_PASSWORD)
        }
        val userDetails = jwtUserDetailsService.loadUserByUsername(loginDto.accountName)
        return jwtTokenUtil.generateToken(userDetails)
    }

    fun getAccountInfo(accountName: String): Account {
        return accountRepository.findByAccountName(accountName)!!
    }

    fun getAccountInfoAll(): List<Account> {
        return accountRepository.findAll()
    }

    fun isAdmin(token: String): Boolean {
        val name = jwtTokenUtil.getUsernameFromToken(token)
        val account = accountRepository.findByAccountName(name)
        if (ObjectUtils.isEmpty(account)) {
            return false
        }
        return account!!.role == "ADMIN"
    }
}